cli/cmd/upgrade.go

```diff
@@ -0,0 +1,207 @@
+	}
+
+	controllerReplicas := uint(1)
+	identityReplicas := uint(1)
+	selector := fmt.Sprintf("%s in (controller, identity)", k8s.ControllerComponentLabel)
+	deploys, err := k.ExtensionsV1beta1().Deployments(controlPlaneNamespace).
+		List(metav1.ListOptions{LabelSelector: selector})
+	if err != nil {
+		return nil, nil, err
+	}
+	for _, deploy := range deploys.Items {
+		if deploy.Spec.Replicas == nil {
+			continue
+		}
+		r := uint(*deploy.Spec.Replicas)
+
+		switch deploy.GetLabels()[k8s.ControllerComponentLabel] {
+		case "controller":
+			controllerReplicas = r
+		case "identity":
+			identityReplicas = r
+		}
+	}
+
+	// Fetch the existing cluster configs or exit.
+	configs := &pb.All{Global: &pb.Global{}, Proxy: &pb.Proxy{}}
+	configMap, err := k.CoreV1().ConfigMaps(controlPlaneNamespace).
+		Get(k8s.ConfigConfigMapName, metav1.GetOptions{})
+	if err != nil {
+		return nil, nil, err
+	}
+	uj := jsonpb.Unmarshaler{}
+	if err := uj.Unmarshal(strings.NewReader(configMap.Data["global"]), configs.Global); err != nil {
+		return nil, nil, err
+	}
+	if err := uj.Unmarshal(strings.NewReader(configMap.Data["proxy"]), configs.Proxy); err != nil {
+		return nil, nil, err
+	}
+
+	// Override the configs from the command-line flags.
+	options.overrideConfigs(configs)
+	j := jsonpb.Marshaler{EmitDefaults: true}
+	globalJSON, err := j.MarshalToString(configs.GetGlobal())
+	if err != nil {
+		return nil, nil, err
+	}
+	proxyJSON, err := j.MarshalToString(configs.GetProxy())
+	if err != nil {
+		return nil, nil, err
+	}
+
+	values := &installValues{
+		// Container images:
+		ControllerImage: fmt.Sprintf("%s/controller:%s", options.dockerRegistry, configs.GetGlobal().GetVersion()),
+		WebImage:        fmt.Sprintf("%s/web:%s", options.dockerRegistry, configs.GetGlobal().GetVersion()),
+		GrafanaImage:    fmt.Sprintf("%s/grafana:%s", options.dockerRegistry, configs.GetGlobal().GetVersion()),
+		PrometheusImage: prometheusImage,
+		ImagePullPolicy: configs.Proxy.ProxyImage.PullPolicy,
+
+		// Kubernetes labels/annotations/resourcse:
+		CreatedByAnnotation:      k8s.CreatedByAnnotation,
+		CliVersion:               k8s.CreatedByAnnotationValue(),
+		ControllerComponentLabel: k8s.ControllerComponentLabel,
+		ProxyContainerName:       k8s.ProxyContainerName,
+		ProxyInjectAnnotation:    k8s.ProxyInjectAnnotation,
+		ProxyInjectDisabled:      k8s.ProxyInjectDisabled,
+
+		// Controller configuration:
+		Namespace:              controlPlaneNamespace,
+		UUID:                   configs.GetGlobal().GetInstallationUuid(),
+		ControllerLogLevel:     options.controllerLogLevel,
+		PrometheusLogLevel:     toPromLogLevel(options.controllerLogLevel),
+		ControllerReplicas:     controllerReplicas,
+		ControllerUID:          options.controllerUID,
+		EnableHA:               options.highAvailability,
+		EnableH2Upgrade:        !options.disableH2Upgrade,
+		NoInitContainer:        configs.GetGlobal().GetCniEnabled(),
```

Comment:
Some of these options (ControllerLogLevel, EnableHA, EnableH2Upgrade) apparently will also have to be added into the Global config to avoid being overwritten during upgrade.

Reply:
I'm _slightly_ uneasy about encoding EnableHA in the config, since that _implies_ a bunch of policy that's inferred elsewhere. It would be ideal if we could just read in the existing resource requests/limits, for instance, instead of overriding.  Same goes for the number of replicas...

We're probably going to have to cut some corners to get upgrade to a useable state for the next edge; so we may want to require that some of these flags be re-set when upgrading...

Alternatively, how would anyone ever _disable_ HA during an upgrade, for instance?

Reply:
I think we should remove `EnableHA` from the `Values` struct and replace it with types corresponding to the various templated values; and then we should satisfy those Values types from config or k8s metadata...  This allows us to at least preserve resource limits/requests and replica counts without storing them in the config (where they are not a source of truth).  Things like EnableH2Upgrade probably need to go into a DestinationContext config...